File: algorithms/aStar.py
import heapq
from algorithms import algorithm
from utils.reader import GraphReader

class AStar(algorithm.Algorithm):
    def __init__(self, graph, start_node, end_node, cities):
        super().__init__(graph, start_node, end_node)
        self.cities = cities
        
        self.closed_set = set()

        self.open_set = [(self._heuristic(start_node, end_node), start_node)]
        
        self.came_from = {}

        self.g_score = {i: float('inf') for i in range(len(graph))}
        self.g_score[start_node] = 0

        self.f_score = {i: float('inf') for i in range(len(graph))}
        self.f_score[start_node] = self._heuristic(start_node, end_node)

    def _heuristic(self, node1, node2):
        return GraphReader.distance(self.cities[node1], self.cities[node2])

    def solve(self):
        self.frontier.add(self.start_node)
        yield self.frontier, self.closed_set, self.start_node, [self.start_node]

        while self.open_set:
            _, current_node = heapq.heappop(self.open_set)
            self.frontier.discard(current_node)

            if current_node == self.end_node:
                yield self.frontier, self.closed_set, current_node, self.reconstruct_path()
                return

            self.closed_set.add(current_node)

            for neighbor in range(len(self.graph[current_node])):
                if self.graph[current_node][neighbor] > 0: # If there is a connection
                    if neighbor in self.closed_set:
                        continue

                    tentative_g_score = self.g_score[current_node] + self.graph[current_node][neighbor]

                    if tentative_g_score < self.g_score[neighbor]:
                        self.came_from[neighbor] = current_node
                        self.g_score[neighbor] = tentative_g_score
                        self.f_score[neighbor] = tentative_g_score + self._heuristic(neighbor, self.end_node)
                        
                        in_open_set = False
                        for _, n in self.open_set:
                            if n == neighbor:
                                in_open_set = True
                                break
                        if not in_open_set:
                            heapq.heappush(self.open_set, (self.f_score[neighbor], neighbor))
                            self.frontier.add(neighbor)
            
            yield self.frontier, self.closed_set, current_node, self.reconstruct_path(current_node)

    def reconstruct_path(self, target_node=None):
        if target_node is None:
            target_node = self.end_node
        path = []
        current_node = target_node
        while current_node in self.came_from:
            path.append(current_node)
            current_node = self.came_from[current_node]
        path.append(self.start_node)
        path.reverse()
        return path

File: src/app/core/algorithms/aStar.ts
import { Algorithm, NodeState } from './algorithm';
import { City } from '../models/city';

export class AStar extends Algorithm {
  private closedSet: Set<number> = new Set();
  private frontierSet: Set<number> = new Set();
  private openSet: { fScore: number, node: number }[] = [];
  private cameFrom: Map<number, number> = new Map();
  private gScore: Map<number, number> = new Map();
  private fScore: Map<number, number> = new Map();
  private cities: City[];

  constructor(graph: number[][], startNode: number, endNode: number, cities: City[]) {
    super(graph, startNode, endNode);
    this.cities = cities;

    for (let i = 0; i < graph.length; i++) {
      this.gScore.set(i, Infinity);
      this.fScore.set(i, Infinity);
    }

    this.gScore.set(startNode, 0);
    this.fScore.set(startNode, this.heuristic(startNode, endNode));
    this.openSet.push({ fScore: this.fScore.get(startNode)!, node: startNode });
  }

  private heuristic(node1: number, node2: number): number {
    const c1 = this.cities[node1];
    const c2 = this.cities[node2];
    return Math.sqrt(Math.pow(c1.x - c2.x, 2) + Math.pow(c1.y - c2.y, 2));
  }

  *solve(): Generator<NodeState, void, unknown> {
    this.frontierSet.add(this.startNode);
    yield {
      frontier: new Set(this.frontierSet),
      visited: new Set(this.closedSet),
      currentNode: this.startNode,
      path: [this.startNode]
    };

    while (this.openSet.length > 0) {
      this.openSet.sort((a, b) => a.fScore - b.fScore);
      const current = this.openSet.shift()!;
      const currentNode = current.node;

      this.frontierSet.delete(currentNode);

      if (currentNode === this.endNode) {
        yield {
          frontier: new Set(this.frontierSet),
          visited: new Set(this.closedSet),
          currentNode: currentNode,
          path: this.reconstructPath()
        };
        return;
      }

      this.closedSet.add(currentNode);

      for (let neighbor = 0; neighbor < this.graph[currentNode].length; neighbor++) {
        const weight = this.graph[currentNode][neighbor];
        if (weight > 0) {
          if (this.closedSet.has(neighbor)) {
            continue;
          }

          const tentativeGScore = this.gScore.get(currentNode)! + weight;

          if (tentativeGScore < this.gScore.get(neighbor)!) {
            this.cameFrom.set(neighbor, currentNode);
            this.gScore.set(neighbor, tentativeGScore);
            const neighborFScore = tentativeGScore + this.heuristic(neighbor, this.endNode);
            this.fScore.set(neighbor, neighborFScore);

            const inOpenSet = this.openSet.find(n => n.node === neighbor);
            if (!inOpenSet) {
              this.openSet.push({ fScore: neighborFScore, node: neighbor });
              this.frontierSet.add(neighbor);
            } else {
              inOpenSet.fScore = neighborFScore;
            }
          }
        }
      }

      yield {
        frontier: new Set(this.frontierSet),
        visited: new Set(this.closedSet),
        currentNode: currentNode,
        path: this.reconstructPath(currentNode)
      };
    }
  }

  private reconstructPath(targetNode: number | null = null): number[] {
    let current = targetNode !== null ? targetNode : this.endNode;
    const path: number[] = [];
    while (this.cameFrom.has(current)) {
      path.push(current);
      current = this.cameFrom.get(current)!;
    }
    path.push(this.startNode);
    return path.reverse();
  }
}
